src/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs

```diff
@@ -330,37 +330,109 @@ public static unsafe SocketError Receive(SafeCloseSocket handle, byte[] buffer,
+            {
+                address = new IPAddress((long)controlBuffer->address);
+            }
 
-            SocketError errorCode = SocketError.Success;
+            return new IPPacketInformation(address, (int)controlBuffer->index);
+        }
+
+        public static unsafe IPPacketInformation GetIPPacketInformation(Interop.Winsock.ControlDataIPv6* controlBuffer)
+        {
+            IPAddress address = IPAddress.IPv6None;
+            if (controlBuffer->length != UIntPtr.Zero)
+            {
+                var addressArray = new byte[Interop.Winsock.IPv6AddressLength];
+                Marshal.Copy((IntPtr)(controlBuffer->address), addressArray, 0, Interop.Winsock.IPv6AddressLength);
+                address = new IPAddress(addressArray);
+            }
+
+            return new IPPacketInformation(address, (int)controlBuffer->index);
+        }
+
+        public static unsafe SocketError ReceiveMessageFrom(Socket socket, SafeCloseSocket handle, byte[] buffer, int offset, int size, ref SocketFlags socketFlags, Internals.SocketAddress socketAddress, out Internals.SocketAddress receiveAddress, out IPPacketInformation ipPacketInformation, out int bytesTransferred)
+        {
+            bool ipv4, ipv6;
+            Socket.GetIPProtocolInformation(socket.AddressFamily, socketAddress, out ipv4, out ipv6);
 
             bytesTransferred = 0;
-            try
+            receiveAddress = socketAddress;
+            ipPacketInformation = default(IPPacketInformation);
+
+            fixed (byte * ptrBuffer = buffer)
+            fixed (byte * ptrSocketAddress = socketAddress.Buffer)
             {
-                // This can throw ObjectDisposedException (retrieving the delegate AND resolving the handle).
-                if (socket.WSARecvMsgBlocking(
-                    handle.DangerousGetHandle(),
-                    Marshal.UnsafeAddrOfPinnedArrayElement(asyncResult._messageBuffer, 0),
-                    out bytesTransferred,
-                    IntPtr.Zero,
-                    IntPtr.Zero) == SocketError.SocketError)
+                Interop.Winsock.WSAMsg wsaMsg;
+                wsaMsg.socketAddress = (IntPtr)ptrSocketAddress;
+                wsaMsg.addressLength = (uint)socketAddress.Size;
+                wsaMsg.flags = socketFlags;
+
+                WSABuffer wsaBuffer;
+                wsaBuffer.Length = size;
+                wsaBuffer.Pointer = (IntPtr)(ptrBuffer + offset);
+                wsaMsg.buffers = (IntPtr)(&wsaBuffer);
+                wsaMsg.count = 1;
+
+                if (ipv4)
                 {
-                    errorCode = GetLastSocketError();
+                    Interop.Winsock.ControlData controlBuffer;
+                    wsaMsg.controlBuffer.Pointer = (IntPtr)(&controlBuffer);
+                    wsaMsg.controlBuffer.Length = sizeof(Interop.Winsock.ControlData);
+
+                    if (socket.WSARecvMsgBlocking(
+                        handle.DangerousGetHandle(),
```

Comment:
Yes.  One of the purposes of SafeHandle is to help ensure that the handle isn't closed while it's still being used.  When a SafeHandle is passed to a DllImport, the marshaling layer does the equivalent of something like:
```C#
bool gotRef = false;
try
{
    safeHandle.DangerousAddRef(ref gotRef);
    IntPtr handle = safeHandle.DangerousGetHandle();
    ... // use handle in call to native function
}
finally
{
    if (gotRef) safeHandle.DangerousRelease();
}
```
When Dispose is called on a SafeHandle, the act of closing the underlying handle is delayed while there's still a ref count outstanding on the SafeHandle.

So if a P/Invoke is written to take a SafeHandle, all of that happens automagically.  If a P/Invoke is instead written to take an IntPtr, that protection is left to the caller to provide, such that somewhere in the call stack you'd expect to see that AddRef/Release wrapping the call.  If that's not happening, and if all call sites are amenable, it'd be best to just change the P/Invoke signature to take the SafeHandle.